common/src/main/java/io/druid/collections/StupidPool.java

```diff
@@ -65,31 +73,60 @@ public StupidPool(
 
     for (int i = 0; i < initCount; i++) {
       objects.add(generator.get());
+      poolSize.incrementAndGet();
     }
   }
 
   public ResourceHolder<T> take()
   {
     final T obj = objects.poll();
-    return obj == null ? new ObjectResourceHolder(generator.get()) : new ObjectResourceHolder(obj);
+    if (obj == null) {
+      return new ObjectResourceHolder(generator.get());
+    } else {
+      poolSize.decrementAndGet();
+      return new ObjectResourceHolder(obj);
+    }
+  }
+
+  /** For tests */
+  long poolSize() {
+    return poolSize.get();
+  }
+
+  private void tryReturnToPool(T object)
+  {
+    long currentPoolSize;
+    do {
+      currentPoolSize = poolSize.get();
+      if (currentPoolSize >= objectsCacheMaxCount) {
+        log.debug("cache num entries is exceeding max limit [%s]", objectsCacheMaxCount);
+        return;
+      }
+    } while (!poolSize.compareAndSet(currentPoolSize, currentPoolSize + 1));
+    if (!objects.offer(object)) {
+      poolSize.decrementAndGet();
+      log.warn(new ISE("Queue offer failed"), "Could not offer object [%s] back into the queue", object);
```

Comment:
Apparently to have a stack trace. I inherited this from the prev. version of this class.